OpenRA.Mods.Common/UpdateRules/Rules/20201213/RenameSupportPowerDescription.cs

```diff
@@ -0,0 +1,32 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2021 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+
+namespace OpenRA.Mods.Common.UpdateRules.Rules
+{
+	public class RenameSupportPowerDescription : UpdateRule
+	{
+		public override string Name => "Support powers now use 'Name' and 'Description' fields like units.";
+		public override string Description => "'Description' was renamed to 'Name' and 'LongDesc' was renamed to 'Description'.";
+
+		public override IEnumerable<string> UpdateActorNode(ModData modData, MiniYamlNode actorNode)
+		{
+			foreach (var traitNode in actorNode.ChildrenContaining("Power"))
```

Comment:
Tbh it would be much cleaner if we used a list of the power names here, but imo this should be good enough... (Unless someone else thinks this isn't?)

Reply:
I want to catch all third party mod support powers which will be catched if they use the same naming convention. I would rather like to cast this into the trait and check if it descents from `SupportPower` but I guess the rules don't allow for that.

Reply:
This will also match unrelated traits like `Power`, `PowerManager`, `InfiltrateForSupportPower`. There shouldn't be any collateral damage in the upstream traits because they don't use `Description` or `LongDesc`, but are we sure that is the case downstream?

Reply:
The current code catches the following traits (when run on D2k):
```
SupportPowerBotModule
SupportPowerManager
PowerManager
GrantConditionOnPowerState
PowerMultiplier
ScalePowerWithHealth
Power
AirstrikePower
WithSupportPowerActivationOverlay
ProduceActorPower
SupportPowerChargeBar
NukePower
```

> I would rather like to cast this into the trait and check if it descents from SupportPower but I guess the rules don't allow for that.

What I can recommend is doing some reflection to that end. For example this:
```cs
readonly Type supportPowerInfoType = typeof(SupportPowerInfo);

HashSet<string> supportPowerInfoDerivedTypes;

public override IEnumerable<string> BeforeUpdate(ModData modData)
{
	supportPowerInfoDerivedTypes = AppDomain.CurrentDomain.GetAssemblies()
		.Where(x => !x.FullName.StartsWith("System"))
		.SelectMany(x => x.GetTypes())
		.Where(x => supportPowerInfoType.IsAssignableFrom(x))
		.Select(x => x.Name)
		.ToHashSet();

	return base.BeforeUpdate(modData);
}

public override IEnumerable<string> UpdateActorNode(ModData modData, MiniYamlNode actorNode)
{
	foreach (var traitNode in actorNode.ChildrenContaining("Power"))
	{
		var traitName = traitNode.Key.Split('@').First() + "Info";
		if (!supportPowerInfoDerivedTypes.Contains(traitName))
			continue;

		traitNode.RenameChildrenMatching("Description", "Name");
		traitNode.RenameChildrenMatching("LongDesc", "Description");
	}

	yield break;
}

```
only catches the following YAML nodes when run on D2k:
```
AirstrikePowerInfo
NukePowerInfo
ProduceActorPowerInfo
```

Reply:
[Discussion on Discord](https://discord.com/channels/153649279762694144/388282819371204608/888756419489312769) suggests this isn't an issue worth the hassle and the code should be good as it is.